File: architecture-skills/design-patterns/examples/visitor/visitor.go
// Visitor pattern in Go.
//
// Classic double dispatch via Accept(). Go's common alternative is a
// type switch (switch s := shape.(type)), which is fine for a handful
// of operations; the visitor interface pays off when many operations
// exist and each must handle every shape (the compiler enforces the
// interface, unlike a type switch's optional default).

package main

import (
	"fmt"
	"math"
)

type ShapeVisitor interface {
	VisitCircle(c Circle)
	VisitRectangle(r Rectangle)
	VisitTriangle(t Triangle)
}

type Shape interface {
	Accept(v ShapeVisitor)
}

type Circle struct {
	Radius float64
}

func (c Circle) Accept(v ShapeVisitor) { v.VisitCircle(c) }

type Rectangle struct {
	Width, Height float64
}

func (r Rectangle) Accept(v ShapeVisitor) { v.VisitRectangle(r) }

type Triangle struct {
	Base, Height float64
}

func (t Triangle) Accept(v ShapeVisitor) { v.VisitTriangle(t) }

type AreaVisitor struct {
	Area float64
}

func (a *AreaVisitor) VisitCircle(c Circle)       { a.Area = math.Pi * c.Radius * c.Radius }
func (a *AreaVisitor) VisitRectangle(r Rectangle) { a.Area = r.Width * r.Height }
func (a *AreaVisitor) VisitTriangle(t Triangle)   { a.Area = 0.5 * t.Base * t.Height }

type SvgExportVisitor struct {
	Markup string
}

func (s *SvgExportVisitor) VisitCircle(c Circle) {
	s.Markup = fmt.Sprintf(`<circle r="%g" />`, c.Radius)
}

func (s *SvgExportVisitor) VisitRectangle(r Rectangle) {
	s.Markup = fmt.Sprintf(`<rect width="%g" height="%g" />`, r.Width, r.Height)
}

func (s *SvgExportVisitor) VisitTriangle(t Triangle) {
	s.Markup = fmt.Sprintf(`<polygon points="0,%g %g,%g 0,0" />`, t.Height, t.Base, t.Height)
}

func main() {
	shapes := []Shape{Circle{2}, Rectangle{3, 4}, Triangle{6, 2}}

	area := &AreaVisitor{}
	svg := &SvgExportVisitor{}
	total := 0.0

	for _, shape := range shapes {
		shape.Accept(area)
		shape.Accept(svg)
		fmt.Printf("area=%6.2f  %s\n", area.Area, svg.Markup)
		total += area.Area
	}
	fmt.Printf("total area: %.2f\n", total)
}

File: architecture-skills/design-patterns/examples/visitor/visitor.rs
// Visitor pattern in Rust.
//
// Idiomatic Rust usually replaces Visitor with an enum + match: the
// compiler enforces exhaustiveness and no accept() plumbing is needed
// (see area_by_match below). The trait-based double-dispatch form is
// worth it when the shape set must stay open to downstream crates.

use std::f64::consts::PI;

struct Circle {
    radius: f64,
}

struct Rectangle {
    width: f64,
    height: f64,
}

struct Triangle {
    base: f64,
    height: f64,
}

trait ShapeVisitor {
    fn visit_circle(&mut self, circle: &Circle);
    fn visit_rectangle(&mut self, rectangle: &Rectangle);
    fn visit_triangle(&mut self, triangle: &Triangle);
}

trait Shape {
    fn accept(&self, visitor: &mut dyn ShapeVisitor);
}

impl Shape for Circle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        visitor.visit_circle(self);
    }
}

impl Shape for Rectangle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        visitor.visit_rectangle(self);
    }
}

impl Shape for Triangle {
    fn accept(&self, visitor: &mut dyn ShapeVisitor) {
        visitor.visit_triangle(self);
    }
}

#[derive(Default)]
struct AreaVisitor {
    area: f64,
}

impl ShapeVisitor for AreaVisitor {
    fn visit_circle(&mut self, c: &Circle) {
        self.area = PI * c.radius * c.radius;
    }
    fn visit_rectangle(&mut self, r: &Rectangle) {
        self.area = r.width * r.height;
    }
    fn visit_triangle(&mut self, t: &Triangle) {
        self.area = 0.5 * t.base * t.height;
    }
}

#[derive(Default)]
struct SvgExportVisitor {
    markup: String,
}

impl ShapeVisitor for SvgExportVisitor {
    fn visit_circle(&mut self, c: &Circle) {
        self.markup = format!("<circle r=\"{}\" />", c.radius);
    }
    fn visit_rectangle(&mut self, r: &Rectangle) {
        self.markup = format!("<rect width=\"{}\" height=\"{}\" />", r.width, r.height);
    }
    fn visit_triangle(&mut self, t: &Triangle) {
        self.markup = format!("<polygon points=\"0,{h} {b},{h} 0,0\" />", h = t.height, b = t.base);
    }
}

// The enum + match alternative most Rust code reaches for first.
enum ShapeEnum {
    Circle { radius: f64 },
}

fn area_by_match(shape: &ShapeEnum) -> f64 {
    match shape {
        ShapeEnum::Circle { radius } => PI * radius * radius,
    }
}

fn main() {
    let shapes: Vec<Box<dyn Shape>> = vec![
        Box::new(Circle { radius: 2.0 }),
        Box::new(Rectangle { width: 3.0, height: 4.0 }),
        Box::new(Triangle { base: 6.0, height: 2.0 }),
    ];

    let mut area = AreaVisitor::default();
    let mut svg = SvgExportVisitor::default();
    let mut total = 0.0;

    for shape in &shapes {
        shape.accept(&mut area);
        shape.accept(&mut svg);
        println!("area={:6.2}  {}", area.area, svg.markup);
        total += area.area;
    }
    println!("total area: {total:.2}");

    let unit = ShapeEnum::Circle { radius: 1.0 };
    println!("enum+match circle area: {:.2}", area_by_match(&unit));
}
